File: ios/Runner/Crypto/CryptoFormat.swift
import Foundation

/// Binary container format shared with Android and the Python reference
/// implementation. See the Kotlin `CryptoFormat.kt` for the authoritative
/// description; kept in sync here.
///
///   [0..5)    "ENCv1"                      5 bytes
<<<<<<< HEAD
///   [5..7)    encryptedKeyLength (uint16BE) 2 bytes  - matches Python's struct.pack(">H", ...)
///   [7..7+N)  RSA-OAEP-SHA256(AES-256 key)  N bytes
=======
///   [5..9)    encryptedKeyLength (uint32BE) 4 bytes
///   [9..9+N)  RSA-OAEP-SHA256(AES-256 key)  N bytes
>>>>>>> dc73e19c0a1ff98a8b3c8cf8e378318f197e1a59
///   [..+12)   GCM IV                        12 bytes
///   [..EOF)   AES-256-GCM ciphertext || 16-byte GCM tag
enum CryptoFormat {
    static let magic = "ENCv1"
    static let magicBytes: [UInt8] = Array(magic.utf8)

    static let gcmIvLength = 12
    static let gcmTagLength = 16
    static let aesKeyLength = 32     // AES-256
    static let rsaKeySizeBits = 2048

    /// Fixed streaming buffer size; memory usage never grows with file size.
    static let streamBufferSize = 8 * 1024 * 1024 // 8 MB
<<<<<<< HEAD
}
=======
}
>>>>>>> dc73e19c0a1ff98a8b3c8cf8e378318f197e1a59

File: android/app/src/main/kotlin/com/bander/sero/crypto/CryptoFormat.kt
package com.bander.sero.crypto

/**
 * Binary container format shared with the Python reference implementation.
 *
 * Layout (all integers big-endian):
 *
 *   [0..5)    "ENCv1"                      5 bytes  - magic / version header
 *   [5..7)    encryptedKeyLength (uint16)  2 bytes  - matches Python's struct.pack(">H", ...)
 *   [7..7+N)  RSA-OAEP-SHA256(AES-256 key) N bytes  - N == encryptedKeyLength
 *   [..+12)   GCM IV                       12 bytes - fresh random per file
 *   [..EOF)   AES-256-GCM ciphertext       ciphertext || 16-byte GCM tag
 *
 * The GCM tag is NOT written separately: both javax.crypto.Cipher (Android)
 * and CommonCrypto's incremental GCM (iOS) append the 16-byte authentication
 * tag to the output of the final encryption call, which is exactly what
 * PyCryptodome / `cryptography` produce on the Python side when the tag is
 * concatenated to the ciphertext. This keeps the three implementations
 * byte-for-byte compatible without any extra framing.
 */
object CryptoFormat {
    const val MAGIC = "ENCv1"
    val MAGIC_BYTES: ByteArray = MAGIC.toByteArray(Charsets.US_ASCII)

    const val GCM_IV_LENGTH = 12          // bytes
    const val GCM_TAG_LENGTH = 16         // bytes
    const val GCM_TAG_LENGTH_BITS = GCM_TAG_LENGTH * 8

    const val AES_KEY_LENGTH = 32         // AES-256
    const val RSA_KEY_SIZE = 2048         // bits, matches typical Python OAEP key size

    /** Size of the streaming read/write buffer. Chosen to keep memory usage
     * constant regardless of file size while still giving good throughput. */
    const val STREAM_BUFFER_SIZE = 8 * 1024 * 1024 // 8 MB

    const val RSA_TRANSFORMATION = "RSA/ECB/OAEPWithSHA-256AndMGF1Padding"
    const val AES_TRANSFORMATION = "AES/GCM/NoPadding"
}